renderc/test/render_test.cc

```diff
@@ -0,0 +1,297 @@
+  ASSERT_TRUE(index.metrics_span.metrics.current_queue().empty());
+
+  // Metric store should no longer keep a reference to the span.
+  ASSERT_EQ(index.metrics_span.size(), 0);
+}
+
+TEST(RenderTest, ManualReference)
+{
+  test::app1::Index index;
+
+  auto span = index.span_with_manual_reference.alloc();
+  ASSERT_TRUE(span.valid());
+
+  auto simple_loc = span.manual_reference().loc();
+
+  // Currently the reference is an invalid reference.
+  ASSERT_EQ(simple_loc, span.manual_reference().invalid);
+
+  // No simple_span is allocated.
+  ASSERT_EQ(index.simple_span.size(), 0);
+
+  {
+    auto s = index.simple_span.alloc();
+    ASSERT_TRUE(s.valid());
+
+    // Save the location of this newly-allocated simple_span.
+    simple_loc = s.loc();
+
+    // Assign it as the reference.
+    span.modify().manual_reference(s.get());
+
+    ASSERT_EQ(index.simple_span.size(), 1);
+    ASSERT_EQ(span.manual_reference().refcount(), 2);
+  }
+
+  ASSERT_TRUE(span.manual_reference().valid());
+  ASSERT_EQ(span.manual_reference().refcount(), 1);
+
+  // It's the same simple_span.
+  ASSERT_EQ(simple_loc, span.manual_reference().loc());
+}
+
+TEST(RenderTest, AutoReference)
+{
+  static constexpr u32 key_one = 11;
+  static constexpr u32 key_two = 22;
+
+  test::app1::Index index;
+
+  auto indexed = index.indexed_span.by_key(key_one);
+  ASSERT_TRUE(indexed.valid());
+  ASSERT_EQ(indexed.number(), key_one);
+
+  // Only one indexed_span exists for now (namely indexed_span{key_one}).
+  ASSERT_EQ(index.indexed_span.size(), 1);
+
+  auto span = index.span_with_auto_reference.alloc();
+  ASSERT_TRUE(span.valid());
+
+  // The auto-reference is not yet valid because the `number` field that is used in the reference key (see test.render)
+  // has not been assigned.
+  ASSERT_FALSE(span.auto_reference().valid());
+
+  // Still only one indexed_span exists.
+  ASSERT_EQ(index.indexed_span.size(), 1);
+
+  // Assign the field that is used in the reference key.
+  span.modify().number(key_two);
+
+  // This caused the reference to be computed and a new indexed_span to be allocated (indexed_span{key_two}).
+  ASSERT_EQ(index.indexed_span.size(), 2);
+
+  // Now the reference is valid.
+  ASSERT_TRUE(span.auto_reference().valid());
+
+  // We have two different indexed_span instances: indexed_span{key_one} and indexed_span{key_two}.
+  ASSERT_NE(indexed.loc(), span.auto_reference().loc());
+  ASSERT_EQ(indexed.number(), key_one);
+  ASSERT_EQ(span.auto_reference().number(), key_two);
+
+  // Set the field that is used in the reference key to the `key_one` value.
+  span.modify().number(key_one);
+
+  // This caused the reference to be recomputed, and now the reference points to indexed_span{key_one}, while the
+  // indexed_span{key_two} instance has been free'd.
+  ASSERT_EQ(index.indexed_span.size(), 1);
+
+  // Those two are the same indexed_span instance (indexed_span{key_one}).
+  ASSERT_EQ(indexed.loc(), span.auto_reference().loc());
+}
```

Comment:
consider also doing `indexed.put()` and check `size()` is 1, to get more verification that `span` actually is holding the reference to the auto-referenced span (and that we didn't bungle the reference counting in the `modify().number()` call.

Reply:
Adressed.